File: race_officer/lib/features/results/scoring/result_code_scoring.dart
import 'package:sailbrowser_flutter/features/race/domain/result_code.dart';

enum ResultCodeAlgorithm {
  na,
  compInSeries,
  compInStartArea,
  avgAll,
  avgBefore,
  scoringPenalty,
  setByHand
}

 sealed class _ResultCodeScoringBase {
  _ResultCodeScoringBase();
}

class ResultCodeScoring extends _ResultCodeScoringBase {

  ResultCodeScoring({
    required this.longSeriesAlgorithm, 
    required this.shortSeriesAlgorithm, 
    required this.isDiscardable, 
    required this.isStartAreaComp,
    required this.isStartedComp, 
    required this.isFinishedComp, 
  });

  final ResultCodeAlgorithm longSeriesAlgorithm; 
  final ResultCodeAlgorithm shortSeriesAlgorithm; 
  final bool isDiscardable;
  final bool isStartAreaComp; 
  final bool isStartedComp; 
  final bool isFinishedComp; 
}

class _ResultCodeScoreLike extends _ResultCodeScoringBase {
  _ResultCodeScoreLike({
    required this.scoreLike,
  });
  final ResultCode scoreLike;
}

/// Returns scoring data for a result code 
ResultCodeScoring? getScoringData(ResultCode code) {
  final resultScoring = _resultCodes[code];

  return switch (resultScoring) {
    _ResultCodeScoreLike() => _resultCodes[resultScoring.scoreLike] as ResultCodeScoring,
    ResultCodeScoring() => resultScoring,
    null => null
  };
}

/// Private map of the algoritms to use for all results codes 
Map<ResultCode, _ResultCodeScoringBase> _resultCodes = {
  ResultCode.notFinished: ResultCodeScoring(
    longSeriesAlgorithm: ResultCodeAlgorithm.na,
    shortSeriesAlgorithm: ResultCodeAlgorithm.na,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: false,
    isFinishedComp: false,
  ),
  ResultCode.ok: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.na,
    longSeriesAlgorithm: ResultCodeAlgorithm.na,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.dnc: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.compInSeries,
    longSeriesAlgorithm: ResultCodeAlgorithm.compInSeries,
    isDiscardable: true,
    isStartAreaComp: false,
    isStartedComp: false,
    isFinishedComp: false,
  ),
  ResultCode.dnf: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.compInSeries,
    longSeriesAlgorithm: ResultCodeAlgorithm.compInStartArea,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: false,
  ),
  ResultCode.ret: _ResultCodeScoreLike(
    scoreLike: ResultCode.dnf,
  ),
  ResultCode.dns: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.compInSeries,
    longSeriesAlgorithm: ResultCodeAlgorithm.compInStartArea,
    isDiscardable: true,
    isStartAreaComp: false,
    isStartedComp: false,
    isFinishedComp: false,
  ),
  ResultCode.ood: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.avgAll,
    longSeriesAlgorithm: ResultCodeAlgorithm.avgAll,
    isDiscardable: true,
    isStartAreaComp: false,
    isStartedComp: false,
    isFinishedComp: false,
  ),
  ResultCode.ocs: _ResultCodeScoreLike(
    scoreLike: ResultCode.dnf,
  ),
  ResultCode.bdf: _ResultCodeScoreLike(
    scoreLike: ResultCode.dnf,
  ),
  ResultCode.dgm: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.compInSeries,
    longSeriesAlgorithm: ResultCodeAlgorithm.compInStartArea,
    isDiscardable: false,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.udf: _ResultCodeScoreLike(
    scoreLike: ResultCode.dnf,
  ),
  ResultCode.zfp: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.scoringPenalty,
    longSeriesAlgorithm: ResultCodeAlgorithm.scoringPenalty,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.dsq: _ResultCodeScoreLike(
    scoreLike: ResultCode.dnf,
  ),
  ResultCode.xpa: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.scoringPenalty,
    longSeriesAlgorithm: ResultCodeAlgorithm.scoringPenalty,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.scp: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.scoringPenalty,
    longSeriesAlgorithm: ResultCodeAlgorithm.scoringPenalty,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.dpi: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.setByHand,
    longSeriesAlgorithm: ResultCodeAlgorithm.setByHand,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.rdg: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.setByHand,
    longSeriesAlgorithm: ResultCodeAlgorithm.setByHand,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.rdga: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.avgAll,
    longSeriesAlgorithm: ResultCodeAlgorithm.avgAll,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.rdgb: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.avgBefore,
    longSeriesAlgorithm: ResultCodeAlgorithm.avgBefore,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  ),
  ResultCode.rdgc: ResultCodeScoring(
    shortSeriesAlgorithm: ResultCodeAlgorithm.setByHand,
    longSeriesAlgorithm: ResultCodeAlgorithm.setByHand,
    isDiscardable: true,
    isStartAreaComp: true,
    isStartedComp: true,
    isFinishedComp: true,
  )
};

File: sailbrowser-web/src/app/scoring/model/result-code-scoring.ts

export const RESULT_CODES = [
  'OK', 'DNC', 'DNS', 'DNF', 'RET', 'OCS', 'BFD', 'NSC', 'UFD', 'DSQ',
  'DNE', 'DGM', 'RDG', 'RDGA', 'RDGB', 'RDGC', 'OOD', 'ZFP',
  'SCP', 'XPA', 'DPI', 'NOT FINISHED'
] as const;

export type ResultCode = (typeof RESULT_CODES)[number];

/**
 * The algorthim to use in calculating the points for a code. 
 */
export enum ResultCodeAlgorithm {
  na = 'na',
  compInSeries = 'compInSeries',      // N + 1 (Total Entries in Series)
  compInStartArea = 'compInStartArea', // N + 1 (Boats that came to the start area)
  avgAll = 'avgAll',                  // Average of all races in series
  avgBefore = 'avgBefore',            // Average of races prior to this one
  scoringPenalty = 'scoringPenalty',  // Numerical finish + % of fleet (RRS 44.3)
  setByHand = 'setByHand'             // Manual point override
}

/**
 * Code Groups
 */
const NOT_IN_START_AREA: ResultCode[] = ['DNC', 'OOD', 'NOT FINISHED'];
const NO_LEGAL_FINISH: ResultCode[] = [
  'DNC', 'DNS', 'DNF', 'RET', 'OCS',
  'BFD', 'UFD', 'DSQ', 'DNE', 'NOT FINISHED',
];
const REDRESS: ResultCode[] = [
  'RDG', 'RDGA', 'RDGB', 'OOD'
];
const NON_DISCARDABLE: ResultCode[] = ['DGM', 'DNE'];

// Whan calculating averages, excluded all codes that dont 
// reflect the competitors performance. Per RRS A9, this should be
// scores from races the boat started and finished.
const FINISHED_AND_SCORED: ResultCode[] = ['OK', 'SCP', 'DPI'];

/**
 * 3. PREDICATES (Public API)
 */
export const isStartAreaComp = (code: ResultCode) => !NOT_IN_START_AREA.includes(code);
export const isFinishedComp = (code: ResultCode) => !NO_LEGAL_FINISH.includes(code);
export const isRedress = (code: ResultCode) => REDRESS.includes(code);
export const isDiscardable = (code: ResultCode) => !NON_DISCARDABLE.includes(code);

export const requiresTime = (code: ResultCode) => FINISHED_AND_SCORED.includes(code);

/**
 * 4. ALGORITHM RESOLVERS (Public API)
 */
export const getShortAlgorithm = (code: ResultCode): ResultCodeAlgorithm => {
  if ((['RDGA', 'OOD'] as ResultCode[]).includes(code)) return ResultCodeAlgorithm.avgAll;
  if (code === 'RDGB') return ResultCodeAlgorithm.avgBefore;
  if ((['ZFP', 'SCP', 'XPA'] as ResultCode[]).includes(code)) return ResultCodeAlgorithm.scoringPenalty;
  if ((['DPI', 'RDG', 'RDGC'] as ResultCode[]).includes(code)) return ResultCodeAlgorithm.setByHand;
  if (code === 'OK') return ResultCodeAlgorithm.na;

  return ResultCodeAlgorithm.compInSeries;
};

export const getLongAlgorithm = (code: ResultCode): ResultCodeAlgorithm => {
  const short = getShortAlgorithm(code);

  // Rule A5.3 (Long Series): DNS/DNF/OCS points based on Starters + 1
  if (short === ResultCodeAlgorithm.compInSeries && isStartAreaComp(code)) {
    return ResultCodeAlgorithm.compInStartArea;
  }
  return short;
};

/** Include result in average pool if RRS Appendix A9 (a) and (b) */
export function includeInAveragePool(code: ResultCode): boolean {
  const algo = getShortAlgorithm(code);
  const isAvgType = [ResultCodeAlgorithm.avgAll, ResultCodeAlgorithm.avgBefore].includes(algo);
  return FINISHED_AND_SCORED.includes(code) && !isAvgType;
}
